README.md
Workarea Afterpay
================================================================================

Afterpay is a pay-by-installments solution. If a user selects afterpay as a payment
option they are taken to the afterpay site where they enter their payment details.
After confirming and placing the order the payment is then captured by the Workarea platform.

Both the US and Australian API endpoints are supported in this implementation. See the **Secrets** section for setting up credentials.
An orders currency code will determine what API endpoints will be used when placing an order. Orders in a non supported currency will not
have Afterpay as an option when checking out.

Refunds are supported in the Workarea Platform.

Workarea Afterpay orders use the following flow:

1. An API call is made on the Checkout Payment step to see if the order is eligible for Afterpay.
2. If eligible an option for Afterpay is displayed.
3. User clicks place order button - an api call is made to get a token.
4. If token creation is successful the token is injected into the DOM and the user is redirected to the Afterpay site.
5. User enters payment details and submits payment.
6. User is taken back to Workarea and payment is authorized or captured.
7. Order confirmation page is displayed.

Afterpay API documentation can be found here: https://docs.afterpay.com/online-api-v2-b857508478e7.html

Implementation Notes
--------------------------------------------------------------------------------

**Product Detail Pages**

This integration makes use of the ```storefront.product_pricing_details``` append point to display the Afterpay pricing on the product detail page. Some custom PDP templates may not have this append point.

If you wish for the Afterpay pricing to appear on your custom PDP template simply add the append point manually by adding the following:

```ruby
= append_partials('storefront.product_pricing_details', product: product)
```


**Testing**

The test API endpoints will be used by default. Production mode can be triggered by setting the ***test*** configuration value in an initializer to ***false***.

```ruby
config.afterpay.test = false
```


**Proxy**

Be sure to whitelist the sandbox and production API endpoints to the Proxy in your hosting environment. Workarea environments running Kubernetes can add this via the command line interface tool.


**Certification Process**

Afterpay has a rigorous certification process that must be completed before using Afterpay in a production environment. This plugin can not guarantee compliance with this process because most of the certification process is branding and marketing specific and is beyond the scope of this integration.

Configuration
--------------------------------------------------------------------------------
The display of Afterpay can be controlled via the ***Afterpay Configuration*** control panel in the Workarea Admin.

The following can be controlled
- Global Display: Turn on/off the display on the entire site, including checkout.
- PDP Display: Turn on/off display on the product detail page.
- Cart Display: Turn on/off display on the cart page


Secrets
--------------------------------------------------------------------------------

The regional US and Australian API endpoints require separate credentials. You can omit regions that are not relevant to your business needs.

    afterpay:
      us:
        merchant_id: YOUR_US_MERCHANT_ID
        secret_key: YOUR_US_SECRET_KEY
      au:
        merchant_id: YOUR_AU_MERCHANT_ID
        secret_key: YOUR_AU_SECRET_KEY


Getting Started
--------------------------------------------------------------------------------

Add the gem to your application's Gemfile:

```ruby
# ...
gem 'workarea-afterpay'
# ...
```

Update your application's bundle.

```bash
cd path/to/application
bundle
```

Workarea Commerce Documentation
--------------------------------------------------------------------------------

See [https://developer.workarea.com](https://developer.workarea.com) for Workarea Commerce documentation.

License
--------------------------------------------------------------------------------

Workarea Afterpay is released under the [Business Software License](LICENSE)


CHANGELOG.md
Workarea Afterpay 2.1.2 (2020-05-06)
--------------------------------------------------------------------------------

*   Fix error showing admin config panel


    Ben Crouse



Workarea Afterpay 2.1.1 (2020-05-04)
--------------------------------------------------------------------------------

*   Respect payment declined in authorize transactions


    Ben Crouse

*   Respect decline payment responses in capture and purchase operations

    Fixes #3

    Ben Crouse

*   Add CI and rubocop linting


    Jeff Yucis



Workarea Afterpay 2.1.0 (2019-11-12)
--------------------------------------------------------------------------------

*   Update Afterpay API version to V2

    New API version allows for a traditional auth then capture
    model of handling payments. Updates also include idempotent
    payment operations
    Jeff Yucis

*   Update Afterpay API version to V2

    New API version allows for a traditional auth then capture
    model of handling payments. Updates also include idempotent
    payment operations
    Jeff Yucis

*   Update README

    Matt Duffy



Workarea Afterpay 2.0.2 (2019-08-21)
--------------------------------------------------------------------------------

*   Open Source!



Workarea Afterpay 2.0.1 (2019-06-11)
--------------------------------------------------------------------------------

*   Show correct dialog for Australia

    AFTERPAY-28
    Jeff Yucis



Workarea Afterpay 2.0.0 (2019-05-28)
--------------------------------------------------------------------------------

*   Afterpay 2.0.0

    no changes from Afterpay v2.0.0.beta.1
    Eric Pigeon



Workarea Afterpay 2.0.0.beta.1 (2019-05-24)
--------------------------------------------------------------------------------

*   Implement afterpay dialog designs

    * Add styleguide component for US and AUS
    * Add afterpay assets - images and raleway font
    * Add afterpay-dialog Sass component
    * Add afterpay_dialog controller
    * Add views for AUS and US dialogs
    * Update afterpay_helper to use afterpay_dialog routes
    * Remove redundant config
    * Display show more link inline with afterpay info in cart & clean up incomplete class
    * Add Helper in engine, remove decorator for application controller

    AFTERPAY-20
    Jake Beresford

*   Integrate with afterpay australia api

    Contextually switch the api endpoints based on a users location, which
    in derived from their currency. Commit also displays the correct
    terms for the users integration type.

    AFTERPAY-16
    Jeff Yucis

*   Remove ID attribute from from_checkout field to prevent duplicate ID errors when installed with Workarea-PayPal

    AFTERPAY-27
    Jake Beresford

*   Pass suburb to billing and shipping addresses

    Split out the billing and shipping address to use the data from
    payment and shipping respectively.

    AFTERPAY-17
    Jeff Yucis

*   Pass test flag to gateway when present in secrets.

    AFTERPAY-18
    Jeff Yucis



Workarea Afterpay 1.0.2 (2019-03-05)
--------------------------------------------------------------------------------

*   Fix casing of Afterpay in translation file

    AFTERPAY-14
    Jeff Yucis



Workarea Afterpay 1.0.1 (2019-02-19)
--------------------------------------------------------------------------------

*   Auto capture afterpay payments

    Afterpay recomends performing an auth with immediate capture when
    processing the payment. This allows refunds to be placed via the
    workarea admin.

    AFTERPAY-12
    Jeff Yucis

*   Clear credit card tender when adding afterpay

    AFTERPAY-13
    Jeff Yucis



Workarea Afterpay 1.0.0 (2018-11-28)
--------------------------------------------------------------------------------

*   Check that the order is in an eligible country

    Commit supresses afterpay from displaying if the currecny is not in a
    supported country.

    AFTERPAY-8
    Jeff Yucis

*   Place afterpay orders immediatly upon returning from afterpay

    AFTERPAY-9
    Jeff Yucis



Workarea Afterpay 1.0.0.beta.3 (2018-10-23)
--------------------------------------------------------------------------------

*   Correct api cache behavior for multisite

    AFTERPAY-7
    Jeff Yucis

*   Improve messaging around tenders on order conf, admin and mailers

    Commit also fixes bug in order balance check for elibility

    AFTERPAY-6
    Jeff Yucis



Workarea Afterpay 1.0.0.beta.2 (2018-10-08)
--------------------------------------------------------------------------------

*   Allow afterpay to be used with split tender orders.

    AFTERPAY-5
    Jeff Yucis



Workarea Afterpay 1.0.0.beta.1 (2018-10-03)
--------------------------------------------------------------------------------

*   Add afterpay as payment option

    Commit allows users to checkout with afterpay. Relevant options are
    displayed if a users order qualifies for afterpay

    AFTERPAY-3
    Jeff Yucis





lib/workarea/afterpay/bogus_gateway.rb
module Workarea
  module Afterpay
    class BogusGateway
      def initialize(*)
      end

      def get_configuration
        b = {
            "minimumAmount": {
                 "amount": "5.00",
                 "currency": "USD"
             },
            "maximumAmount": {
                "amount": "500.00",
                "currency": "USD"
            }
          }

        Response.new(response(b))
      end

      def get_order(token, options = {})
        b = {
              "billing": {
                  "countryCode": "US",
                  "line1": "29 fort mott road",
                  "name": "Jeffrey Yucis",
                  "phoneNumber": "3027507743",
                  "postcode": "08070",
                  "state": "NJ",
                  "suburb": "pennsville"
              },
              "consumer": {
                  "email": "jyucis@gmail.com",
                  "givenNames": "Jeffrey",
                  "phoneNumber": "3027507743",
                  "surname": "Yucis"
              },
              "courier": {},
              "discounts": [],
              "expires": "2019-04-02T16:52:43.920Z",
              "items": [
                  {
                      "name": "Small Wool Coat",
                      "price": {
                          "amount": "89.30",
                          "currency": "USD"
                      },
                      "quantity": 1,
                      "sku": "139364945-9"
                  }
              ],
              "merchant": {
                  "redirectCancelUrl": "http://localhost:3000/afterpay/cancel",
                  "redirectConfirmUrl": "http://localhost:3000/afterpay/complete"
              },
              "merchantReference": "63837A272F",
              "paymentType": "PAY_BY_INSTALLMENT",
              "shipping": {
                  "countryCode": "US",
                  "line1": "29 fort mott road",
                  "name": "Jeffrey Yucis",
                  "phoneNumber": "3027507743",
                  "postcode": "08070",
                  "state": "NJ",
                  "suburb": "pennsville"
              },
              "shippingAmount": {
                  "amount": "6.00",
                  "currency": "USD"
              },
              "taxAmount": {
                  "amount": "0.00",
                  "currency": "USD"
              },
              "token": "9tlqhfgebl6mu2g9t98rre2ia25ri2hvadc4aaimvpca1p9fma5j",
              "amount": {
                  "amount": "95.30",
                  "currency": "USD"
              }
          }
        Response.new(response(b))
      end

      def create_order(order)
        b = {
          "token": "q54l9qd907m6iqqqlcrm5tpbjjsnfo47vsm59gqrfnd2rqefk9hu",
          "expires": "2016-05-10T13:14:01Z"
        }

        Response.new(response(b))
      end

      def capture(payment_id, amount, request_id)
        if payment_id == 'declined'
          Response.new(response(payment_response_body(status: 'DECLINED'), 200))
        else
          Response.new(response(payment_response_body, 200))
        end
      end

      def authorize(token, order_id = "", request_id)
        if token == "error_token"
          Response.new(response(capture_error_response_body, 402))
        elsif token == "timeout_token"
          Response.new(response(nil, 502))
        elsif token == 'declined'
          Response.new(response(payment_response_body(status: 'DECLINED'), 200))
        else
          Response.new(response(payment_response_body, 200))
        end
      end

      def purchase(token, order_id = "", request_id)
        if token == "error_token"
          Response.new(response(capture_error_response_body, 402))
        elsif token == 'declined'
          Response.new(response(payment_response_body(status: 'DECLINED'), 200))
        else
          Response.new(response(payment_response_body, 200))
        end
      end

      def refund(afterpay_order_id, amount, request_id)
        b = {
          "requestId": "61cdad2d-8e10-42ec-a97b-8712dd7a8ca9",
          "amount": {
              "amount": "10.00",
              "currency": "USD"
          },
          "merchantReference": "RF127261AD22",
          "refundId": "56785678",
          "refundedAt": "2015-07-10T15:01:14.123Z"
        }

        Response.new(response(b))
      end

      def void(payment_id)
        Response.new(response(payment_response_body))
      end

      private

        def response(body, status = 200)
          response = Faraday.new do |builder|
            builder.adapter :test do |stub|
              stub.get("/v1/bogus") { |env| [ status, {}, body.to_json ] }
            end
          end
          response.get("/v1/bogus")
        end

        def capture_error_response_body
          {
            "errorCode": "declined",
            "errorId": "c4d64cbc3e61a26f",
            "message": "Payment declined",
            "httpStatusCode":  402
          }
        end

        def payment_response_body(status: 'APPROVED')
          {
            "id": "12345678",
             "token": "q54l9qd907m6iqqqlcrm5tpbjjsnfo47vsm59gqrfnd2rqefk9hu",
             "status": status,
             "created": "2015-07-14T10:08:14.123Z",
             "totalAmount": {
                "amount": "21.85",
                "currency": "USD"
             },
             "merchantReference": "merchantOrder-1234",
             "refunds": [],
             "orderDetails": {
                  "consumer": {
                     "phoneNumber": "0200000000",
                     "givenNames": "Joe",
                     "surname": "Consumer",
                     "email": "test@afterpay.com"
                  },
                 "billing": {
                     "name": "Joe Consumer",
                     "line1": "75 Queen St",
                     "state": "New York",
                     "postcode": "10001",
                     "countryCode": "US",
                     "phoneNumber": "2120000000"
                 },
                 "shipping": {
                     "name": "Joe Consumer",
                     "line1": "75 Queen St",
                     "state": "New York",
                     "postcode": "10001",
                     "countryCode": "US",
                     "phoneNumber": "2120000000"
                 },
                 "courier": {},
                 "items": [
                    {
                        "name": "widget",
                        "sku": "12341234",
                        "quantity": 1,
                        "price": {
                            "amount": "10.00",
                            "currency": "USD"
                        }
                    }
                 ],
                 "discounts": [
                    {
                        "displayName": "10% Off Subtotal",
                        "amount": {
                            "amount": "1.00",
                            "currency": "USD"
                        }
                    }
                 ],
                 "shippingAmount": {
                    "amount": "10.00",
                    "currency": "USD"
                 },
                 "taxAmount": {
                    "amount": "2.85",
                    "currency": "USD"
                 }
             },
            "events": []
          }
        end
    end
  end
end
